### PR TITLE
Request constraints

### LINKED ISSUE #997: Add required filter support at table level

### PR DESCRIPTION
**Pull request by kevinhinterlong**
Resolves #997 

## Description
Adds support for basic metadata driven constraints at the table level such as specifying a date time filter for fili tables
```js
{
    property: 'filters',
    matches: {
      type: 'timeDimension',
      field: 'tableName.dateTime'
    }
}
```

or a date time column for fili tables that do not support an all grain

```js
{
    property: 'columns',
    matches: {
      type: 'timeDimension',
      field: 'tableName.dateTime'
    }
}
```

## Proposed Changes

- metadata drive table constraints
- basic constraint solver that adds missing filters/columns
- consumer to listen for switching table/new report events and re-add constraints if needed
- metric/dimension selector and column config now always add new filters instead of toggling

## Screenshots
required filters/columns

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by cythrawll — APPROVED** [review]
Small suggestions.

File: packages/data/addon/models/metadata/request-constraint.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  name!: string;
+  description?: string | undefined;
+  type!: 'existence';
+  constraint!: ExistenceConstraint;
+  source!: string;
+
+  isSatisfied(request: RequestV2): boolean {
+    const { property, matches: partialProperties } = this.constraint;
+    const func = matches(partialProperties);
+    //@ts-ignore
```
[the remark this answers is not in the document]
Comment (kevinhinterlong) [reply packages/data/addon/models/metadata/request-constraint.ts]: This was ignoring an issue where `.filter` tried to resolve for the type `Column[] | Filter[]` and the signatures don't match (whereas  `(Column | Filter)[]` would match) but apparently `some` does not have this issue

File: packages/reports/addon/consumers/request/constraint.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright 2020, Yahoo Holdings Inc.
```
Comment (jkusa) [inline packages/reports/addon/consumers/request/constraint.ts:2]: 2021

File: packages/reports/tests/acceptance/column-config-test.js [hunk 3]
```diff
@@ -362,6 +362,20 @@ module('Acceptance | Navi Report | Column Config', function (hooks) {
...
+    assert.dom('.filter-builder__subject').exists({ count: 1 }, 'There is 1 filter to start');
+    await clickItem('dimension', 'Age');
+
+    await click('.navi-column-config-base__filter-icon');
+    assert.dom('.filter-builder__subject').exists({ count: 2 }, 'Clicking the filter button adds a new filter');
+
+    await click('.navi-column-config-base__filter-icon');
+    assert.dom('.filter-builder__subject').exists({ count: 3 }, 'Clicking the filter button again adds a new filter');
+  });
+
```
Comment (jkusa) [inline packages/reports/tests/acceptance/column-config-test.js:378]: What is the status on the test below?

File: packages/reports/addon/services/request-constrainer.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    const serialized = request.serialize() as RequestV2;
+    const unsatisfiedConstraints =
+      request.tableMetadata?.requestConstraints.filter((c) => !c.isSatisfied(serialized)) || [];
+    if (unsatisfiedConstraints.length > 0) {
+      this.naviNotifications.add({
+        title: `The following requirements are not satisfied`,
+        context: unsatisfiedConstraints.map((c) => c.description).join('; '),
+        style: 'warning',
+        timeout: 'none',
```
[the remark this answers is not in the document]
Comment (kevinhinterlong) [reply packages/reports/addon/services/request-constrainer.ts:78]: Yes right now it should be impossible to make a constraint that can't be solved so I think it's fair that this needs to be acknowledged

File: packages/reports/tests/acceptance/column-config-test.js (hunk 3 above)
Comment (kevinhinterlong) [reply packages/reports/tests/acceptance/column-config-test.js:378]: Probably mostly covered by other tests but I fixed it and reenabled it for now

File: packages/reports/addon/templates/components/navi-column-config/item.hbs [hunk 5]
```diff
@@ -10,8 +10,10 @@
     <button
       type="button"
       aria-label="delete {{dasherize @column.fragment.type}} {{@column.fragment.displayName}}"
-      class="navi-column-config-item__remove-icon"
-      {{on "click" @onRemoveColumn}}>
+      class="navi-column-config-item__remove-icon link"
```
Comment (jkusa) [inline packages/reports/addon/templates/components/navi-column-config/item.hbs:13]:
```suggestion
      class="navi-column-config-item__remove-icon link is-secondary"
```
I think will align the colors

**Review by jkusa — APPROVED** [review]
Just the comment about the filter colors, otherwise 👍

**Review by jkusa — APPROVED** [review]

**Pull request merged**